## Работа с файлами

In [ ]:
from google.colab import drive
import os
drive.mount('/content/gdrive')
path = '/content/gdrive/MyDrive/Python/Data_Science_8/data_5'
data_acts = os.listdir(path)
print(data_acts)

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).
['walking', 'stairs', 'idle', 'running']


In [ ]:
for act in data_acts:
  pathd = os.path.join(path, act)
  files = os.listdir(pathd)
  print(act, '-   ', len(files))

walking -    1850
stairs -    165
idle -    1039
running -    3408


Получили количество файлов с данными - мы не знаем где это данные с одного человека или с разных и какие конкретно данные - все перемешано - с разных людей типичные активности собраны в файл по 30 примеров. и из множества этих файлов собран датасет. можно заметить что файлов с примерами для движения по ступенькам-  менее всего)))  наверное тяжело найти было спортивных людей.....

## Посмотрим на структуру файла

In [ ]:
import pandas as pd
import numpy as np
d_files = os.listdir(os.path.join(path,'idle'))
d_file = os.path.join(os.path.join(path,'idle'), d_files[0])
data_fr = pd.read_csv(d_file)
data_fr['activity'] = 0
#print(data_fr)

## Формируем датасет для активности idle, кодируемой числом 0

In [ ]:
d_files = os.listdir(os.path.join(path,'idle'))   # список файлов в папке
data_fr_0 = pd.DataFrame([])
for file_i in d_files:
  d_file = os.path.join(os.path.join(path,'idle'), file_i)
  d_fr = pd.read_csv(d_file)
  d_fr['activity'] = 0        # добавляем столбец - будущий y
  data_fr_0 = pd.concat([data_fr_0, d_fr], axis=0, ignore_index=True)  # соединяемм датасеты из файлов в один датасет
#print(data_fr_0.head(35))
#print(data_fr_0.tail(35))
if len(d_files) * 30 == data_fr_0.shape[0]:  ## проверка корректности датасета по длине
  print('Dataset IDLE convert correct')
else:
  print('Dataset IDLE conver UNcorrect - you should find a bag')

Dataset IDLE convert correct


## Формируем датасет для активности stairs, кодируемой числом 1

In [ ]:
d_files = os.listdir(os.path.join(path,'stairs'))   # список файлов в папке
data_fr_1 = pd.DataFrame([])
for file_i in d_files:
  d_file = os.path.join(os.path.join(path,'stairs'), file_i)
  d_fr = pd.read_csv(d_file)
  d_fr['activity'] = 1        # добавляем столбец - будущий y
  data_fr_1 = pd.concat([data_fr_1, d_fr], axis=0, ignore_index=True)  # соединяемм датасеты из файлов в один датасет
#print(data_fr_1.head(35))
#print(data_fr_1.tail(35))
if len(d_files) * 30 == data_fr_1.shape[0]:  ## проверка корректности датасета по длине
  print('Dataset STAIRS convert correct')
else:
  print('Dataset STAIRS conver UNcorrect - you should find a bag')

Dataset STAIRS convert correct


## Формируем датасет для активности running, кодируемой числом 2

In [ ]:
d_files = os.listdir(os.path.join(path,'running'))   # список файлов в папке
data_fr_2 = pd.DataFrame([])
for file_i in d_files:
  d_file = os.path.join(os.path.join(path,'running'), file_i)
  d_fr = pd.read_csv(d_file)
  d_fr['activity'] = 2        # добавляем столбец - будущий y
  data_fr_2 = pd.concat([data_fr_2, d_fr], axis=0, ignore_index=True)  # соединяемм датасеты из файлов в один датасет
#print(data_fr_2.head(35))
#print(data_fr_2.tail(35))
if len(d_files) * 30 == data_fr_2.shape[0]:  ## проверка корректности датасета по длине
  print('Dataset STAIRS convert correct')
else:
  print('Dataset STAIRS conver UNcorrect - you should find a bag')

Dataset STAIRS convert correct


## Формируем датасет для активности walking, кодируемой числом 3

In [ ]:
d_files = os.listdir(os.path.join(path,'walking'))   # список файлов в папке
data_fr_3 = pd.DataFrame([])
for file_i in d_files:
  d_file = os.path.join(os.path.join(path,'walking'), file_i)
  d_fr = pd.read_csv(d_file)
  d_fr['activity'] = 3        # добавляем столбец - будущий y
  data_fr_3 = pd.concat([data_fr_3, d_fr], axis=0, ignore_index=True)  # соединяемм датасеты из файлов в один датасет
#print(data_fr_3.head(35))
#print(data_fr_3.tail(35))
if len(d_files) * 30 == data_fr_3.shape[0]:  ## проверка корректности датасета по длине
  print('Dataset STAIRS convert correct')
else:
  print('Dataset STAIRS conver UNcorrect - you should find a bag')

Dataset STAIRS convert correct


## формируем общий датасет

In [ ]:
dataset_target = pd.concat([data_fr_0, data_fr_1], axis=0, ignore_index=True)
dataset_target = pd.concat([dataset_target, data_fr_2], axis=0, ignore_index=True)
dataset_target = pd.concat([dataset_target, data_fr_3], axis=0, ignore_index=True)
print(dataset_target.shape)
print(dataset_target.columns)

(193860, 4)
Index(['accelerometer_X', 'accelerometer_Y', 'accelerometer_Z', 'activity'], dtype='object')


## Готовим данные для обучения классификатора

In [ ]:
from sklearn.utils import shuffle
from sklearn import model_selection
df_shuff = shuffle(dataset_target, random_state=42)   # перемешаем датасет для валидности
X = np.array(df_shuff[['accelerometer_X', 'accelerometer_Y', 'accelerometer_Z']])  # матрица Х - пока пробуем без стандартизации данных
y = np.array(df_shuff['activity'])                                                 # вектор у
X_train, X_test, y_train, y_test = model_selection.train_test_split(X, y, train_size=0.25) # разбили для оценки скора после обучения

# SVC классификатор - по сырым данным

In [ ]:
from sklearn.svm import SVC
clf = SVC(C=1.0, decision_function_shape='ovr') # one vs rest
clf.fit(X_train, y_train)   # обучили классификатор
print(clf.predict([[1.0, 1.0, 1.0]]))   # тестовое предсказание
print(clf.score(X_test, y_test))

[2]
0.8900718731730802


имеем точность 0.89 - єто немного
проверим для стратегии - ovo



In [ ]:
clf = SVC(C=1.0, decision_function_shape='ovo') # one vs one
clf.fit(X_train, y_train)   # обучили классификатор
print(clf.predict([[1.0, 1.0, 1.0]]))   # тестовое предсказание
print(clf.score(X_test, y_test))

[2]
0.8900718731730802


Видим что стратегия не особо улучшает точность модели. Попробуем с коэффициентом регуляризации

In [ ]:
clf = SVC(C=0.5, decision_function_shape='ovr') # one vs rest С=0.5
clf.fit(X_train, y_train)   # обучили классификатор
print(clf.predict([[1.0, 1.0, 1.0]]))   # тестовое предсказание
print(clf.score(X_test, y_test))

[2]
0.8890539564634272


In [ ]:
clf = SVC(C=2.0, decision_function_shape='ovr') # one vs rest С=2.0
clf.fit(X_train, y_train)   # обучили классификатор
print(clf.predict([[1.0, 1.0, 1.0]]))   # тестовое предсказание
print(clf.score(X_test, y_test))

[2]
0.8935451700539908


In [ ]:
clf = SVC(C=10.0, decision_function_shape='ovr') # one vs rest С=10.0
clf.fit(X_train, y_train)   # обучили классификатор
print(clf.predict([[1.0, 1.0, 1.0]]))   # тестовое предсказание
print(clf.score(X_test, y_test))

[2]
0.8947350321537879


Итак - при сырых НАШИХ!!!  данных модель почти не меняет точность при смене стратегии, но увеличивает точность при увеличении коэффициента регуляризации

Попробуем тоже самое при нормализованых данных

In [19]:
from sklearn import preprocessing
X_norm = preprocessing.normalize(X, norm='l2')
X_train, X_test, y_train, y_test = model_selection.train_test_split(X_norm, y, train_size=0.25) # разбили для оценки скора после обучения

повторяем те же модели для нормализованых данных

In [20]:
clf = SVC(C=1.0, decision_function_shape='ovr') # one vs rest
clf.fit(X_train, y_train)   # обучили классификатор
print(clf.predict([[1.0, 1.0, 1.0]]))   # тестовое предсказание
print(clf.score(X_test, y_test))

[2]
0.8807661886584821


In [21]:
clf = SVC(C=1.0, decision_function_shape='ovo') # one vs one
clf.fit(X_train, y_train)   # обучили классификатор
print(clf.predict([[1.0, 1.0, 1.0]]))   # тестовое предсказание
print(clf.score(X_test, y_test))

[2]
0.8807661886584821


In [22]:
clf = SVC(C=0.5, decision_function_shape='ovr') # one vs rest С=0.5
clf.fit(X_train, y_train)   # обучили классификатор
print(clf.predict([[1.0, 1.0, 1.0]]))   # тестовое предсказание
print(clf.score(X_test, y_test))

[2]
0.8794456480621754


In [23]:
clf = SVC(C=2.0, decision_function_shape='ovr') # one vs rest С=2.0
clf.fit(X_train, y_train)   # обучили классификатор
print(clf.predict([[1.0, 1.0, 1.0]]))   # тестовое предсказание
print(clf.score(X_test, y_test))

[2]
0.8819216616802503


In [24]:
clf = SVC(C=10.0, decision_function_shape='ovr') # one vs rest С=10.0
clf.fit(X_train, y_train)   # обучили классификатор
print(clf.predict([[1.0, 1.0, 1.0]]))   # тестовое предсказание
print(clf.score(X_test, y_test))

[2]
0.8833660029574607


Пробуем все то же самое но модель - RandomForest

In [25]:
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier(max_depth=4, random_state=0, bootstrap=True) # bootstrap data
clf.fit(X_train, y_train)   # обучили классификатор
print(clf.predict([[1.0, 1.0, 1.0]]))   # тестовое предсказание
print(clf.score(X_test, y_test))

[2]
0.8793012139344544


In [27]:
clf = RandomForestClassifier(max_depth=4, random_state=0, bootstrap=False) # not bootstrap data
clf.fit(X_train, y_train)   # обучили классификатор
print(clf.predict([[1.0, 1.0, 1.0]]))   # тестовое предсказание
print(clf.score(X_test, y_test))

[2]
0.8798583169985212


In [28]:
clf = RandomForestClassifier(max_depth=10, random_state=0, bootstrap=True) # bootstrap data
clf.fit(X_train, y_train)   # обучили классификатор
print(clf.predict([[1.0, 1.0, 1.0]]))   # тестовое предсказание
print(clf.score(X_test, y_test))

[2]
0.9177825922487018


In [29]:
clf = RandomForestClassifier(max_depth=20, random_state=0, bootstrap=True) # bootstrap data
clf.fit(X_train, y_train)   # обучили классификатор
print(clf.predict([[1.0, 1.0, 1.0]]))   # тестовое предсказание
print(clf.score(X_test, y_test))

[2]
0.994442724990543


Видим модель RandomForest работает намного быстрее чем SVC и точнее при оответствующей глубине построения дерева. В общем то именно это и написано в мануале, что SVC сильно замедляется если величина семплов превышает десятки тысяч, а у нас 193860 семплов.
Вывод - модель RandomForest предпочтительней при больших датасетах